File: nand2tetris/jack-compiler/cpp/symboltable.cpp

#include "symbol.h"
#include "symboltable.h"
#include <iostream>
#include <string>

using namespace std;
using namespace sym;

// Constructor
SymbolTable::SymbolTable()
{
  class_scope_.clear();
  subroutine_scope_.clear();
  for (int i = 0; i < 4; i++)
  {
    count[i] = 0;
  }
}

void SymbolTable::startSubroutine(void)
{
  subroutine_scope_.clear();
  count[sym::ARG] = 0;
  count[sym::VAR] = 0;
}

void SymbolTable::define(string name, string type, sym::Kind kind)
{
  Symbol s = {name, type, kind, count[kind]};

  if (kind == sym::STATIC or kind == sym::FIELD)
  {
    class_scope_[name] = s;
  }
  else if (kind == sym::ARG or kind == sym::VAR)
  {
    subroutine_scope_[name] = s;
  }

  count[kind]++;
}

int SymbolTable::varCount(sym::Kind kind) { return count[kind]; }

bool hasKey(map<string, Symbol> &scope, string key)
{
  return !(scope.find(key) == scope.end());
}

Symbol SymbolTable::resolve(string name)
{
  Symbol s;
  if (resolve(name, s))
  {
    return s;
  }
  else
  {
    throw string("Variable not defined!");
  }
}

bool SymbolTable::resolve(string name, Symbol &s)
{
  bool r = true;
  if (hasKey(subroutine_scope_, name))
    s = subroutine_scope_[name];
  else if (hasKey(class_scope_, name))
    s = class_scope_[name];
  else
  {
    r = false;
  }

  return r;
}

File: nand2tetris/jack-compiler/scala/src/main/scala/SymbolTable.scala

package jackcompiler

import scala.collection.mutable

object Kind extends Enumeration {
    type Kind = Value
    val STATIC, FIELD, ARG, VAR = Value
}



case class Symbol(name: String, typeOf: String, kind: Kind.Kind, index: Int)

class SymbolTable {
  

  private val classScope: mutable.Map[String, Symbol] = mutable.Map()
  private val subroutineScope: mutable.Map[String, Symbol] = mutable.Map()
  private val countVars: mutable.Map[Kind.Kind, Int] = mutable.Map(
    Kind.ARG -> 0,
    Kind.VAR -> 0,
    Kind.STATIC -> 0,
    Kind.FIELD -> 0
  )

  def startSubroutine(): Unit = {
    subroutineScope.clear()
    countVars.update(Kind.ARG, 0)
    countVars.update(Kind.VAR, 0)
  }

  private def scope(kind: Kind.Kind): mutable.Map[String, Symbol] =
    if (kind == Kind.STATIC || kind == Kind.FIELD) classScope else subroutineScope

  def define(name: String, typeOf: String, kind: Kind.Kind): Unit = {
    val scopeTable = scope(kind)
    if (scopeTable.contains(name)) {
      throw new RuntimeException("Variable already defined")
    }

    val s = Symbol(name, typeOf, kind, varCount(kind))
    scopeTable(name) = s
    countVars.update(kind, countVars(kind) + 1)
  }

  def resolve(name: String): Option[Symbol] = {
    subroutineScope.get(name).orElse(classScope.get(name))
  }

  def varCount(kind: Kind.Kind): Int = {
    countVars(kind)
  }
}
